File: src/test/java/network/crypta/fs/AppDirsTest.java
package network.crypta.fs;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;
import network.crypta.config.CryptadConfig;
import network.crypta.support.SimpleFieldSet;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

@SuppressWarnings("java:S100") // Enforce method_whenCondition_expectOutcome naming for tests.
class AppDirsTest {

  @TempDir Path tmp;

  private static String norm(String s) {
    return s.replace('\\', '/');
  }

  private static void ensureDir(Path dir) {
    try {
      Files.createDirectories(dir);
    } catch (IOException e) {
      throw new AssertionError("Failed to create test directory: " + dir, e);
    }
  }

  private Map<String, String> sysProps(Path home, Path tmpdir) {
    Map<String, String> p = new HashMap<>();
    p.put("user.home", home.toString());
    p.put("java.io.tmpdir", tmpdir.toString());
    p.put("os.name", "Linux");
    return p;
  }

  @Test
  void resolve_whenLinuxXdgUnset_expectDefaultsUnderHome() {
    // Arrange
    Path home = tmp.resolve("home");
    Path t = tmp.resolve("t");
    ensureDir(home);
    ensureDir(t);
    Map<String, String> env = new HashMap<>();
    AppEnv ae = new AppEnv(env, "Linux", "tester");
    AppDirs dirs = new AppDirs(env, sysProps(home, t), new HashMap<>(), ae);

    // Act
    Resolved r = dirs.resolve();

    // Assert
    assertTrue(norm(r.configDir().toString()).contains(".config/cryptad/config"));
    assertTrue(norm(r.dataDir().toString()).contains(".local/share/cryptad/data"));
    assertTrue(Files.exists(r.configDir()));
  }

  @Test
  void resolve_whenLinuxXdgSet_expectEnvBasesUsed() {
    // Arrange
    Path root = tmp;
    Path home = root.resolve("home");
    Path t = root.resolve("t");
    Path xdgConfig = root.resolve("xdg-config");
    Path xdgData = root.resolve("xdg-data");
    Path xdgCache = root.resolve("xdg-cache");
    ensureDir(home);
    ensureDir(t);
    ensureDir(xdgConfig);
    ensureDir(xdgData);
    ensureDir(xdgCache);
    Map<String, String> env = new HashMap<>();
    env.put("XDG_CONFIG_HOME", xdgConfig.toString());
    env.put("XDG_DATA_HOME", xdgData.toString());
    env.put("XDG_CACHE_HOME", xdgCache.toString());
    AppEnv ae = new AppEnv(env, "Linux", "tester");
    AppDirs dirs = new AppDirs(env, sysProps(home, t), new HashMap<>(), ae);

    // Act
    Resolved r = dirs.resolve();

    // Assert
    assertTrue(r.configDir().startsWith(xdgConfig));
    assertTrue(r.dataDir().startsWith(xdgData));
    assertTrue(r.cacheDir().startsWith(xdgCache));
  }

  @Test
  void resolve_whenMacNativeWithoutXdg_expectLibraryPaths() {
    // Arrange
    Path home = tmp.resolve("home");
    ensureDir(home);
    Map<String, String> env = new HashMap<>();
    AppEnv ae = new AppEnv(env, "Mac OS X", "user");
    Map<String, String> sp = sysProps(home, tmp);
    sp.put("os.name", "Mac OS X");
    AppDirs dirs = new AppDirs(env, sp, new HashMap<>(), ae);

    // Act
    Resolved r = dirs.resolve();

    // Assert
    assertTrue(
        norm(r.configDir().toString()).contains("Library/Application Support/Cryptad/config"));
    assertTrue(norm(r.cacheDir().toString()).contains("Library/Caches/Cryptad"));
  }

  @Test
  void resolve_whenMacWithXdg_expectXdgPaths() {
    // Arrange
    Path home = tmp.resolve("home");
    Path xdgConfig = tmp.resolve("xdg");
    ensureDir(home);
    ensureDir(xdgConfig);
    Map<String, String> env = new HashMap<>();
    env.put("XDG_CONFIG_HOME", xdgConfig.toString());
    AppEnv ae = new AppEnv(env, "Mac OS X", "user");
    Map<String, String> sp = sysProps(home, tmp);
    sp.put("os.name", "Mac OS X");
    AppDirs dirs = new AppDirs(env, sp, new HashMap<>(), ae);

    // Act
    Resolved r = dirs.resolve();

    // Assert
    assertTrue(r.configDir().startsWith(xdgConfig));
  }

  @Test
  void resolve_whenSnapWithUserCommon_expectDataUnderCommon() {
    // Arrange
    Path root = tmp;
    Path home = root.resolve("home");
    Path common = root.resolve("snap-common");
    ensureDir(home);
    ensureDir(common);
    Map<String, String> env = new HashMap<>();
    env.put("SNAP", "/snap/app");
    env.put("SNAP_USER_COMMON", common.toString());
    env.put("SNAP_USER_DATA", home.toString());
    env.put("XDG_CONFIG_HOME", home.resolve(".config").toString());
    env.put("XDG_DATA_HOME", home.resolve(".local/share").toString());
    env.put("XDG_CACHE_HOME", home.resolve(".cache").toString());
    AppEnv ae = new AppEnv(env, "Linux", "user");
    AppDirs dirs = new AppDirs(env, sysProps(home, root), new HashMap<>(), ae);

    // Act
    Resolved r = dirs.resolve();

    // Assert
    assertTrue(r.dataDir().startsWith(common.resolve("cryptad")));
  }

  @Test
  void resolve_whenSnapWithoutUserCommon_expectXdgAndRuntimeDir() {
    // Arrange
    Path root = tmp;
    Path home = root.resolve("home");
    Path xdgConfig = root.resolve("xdg-config");
    Path xdgData = root.resolve("xdg-data");
    Path xdgCache = root.resolve("xdg-cache");
    Path xdgRt = root.resolve("xdg-rt");
    ensureDir(home);
    ensureDir(xdgConfig);
    ensureDir(xdgData);
    ensureDir(xdgCache);
    ensureDir(xdgRt);
    Map<String, String> env = new HashMap<>();
    env.put("SNAP", "/snap/app");
    env.put("XDG_CONFIG_HOME", xdgConfig.toString());
    env.put("XDG_DATA_HOME", xdgData.toString());
    env.put("XDG_CACHE_HOME", xdgCache.toString());
    env.put("XDG_RUNTIME_DIR", xdgRt.toString());
    AppEnv ae = new AppEnv(env, "Linux", "user");
    AppDirs dirs = new AppDirs(env, sysProps(home, root), new HashMap<>(), ae);

    // Act
    Resolved r = dirs.resolve();

    // Assert
    assertTrue(r.configDir().startsWith(xdgConfig.resolve("cryptad/config")));
    assertTrue(r.dataDir().startsWith(xdgData.resolve("cryptad/data")));
    assertTrue(r.cacheDir().startsWith(xdgCache.resolve("cryptad")));
    assertTrue(r.runDir().startsWith(xdgRt.resolve(network.crypta.fs.Dirs.APP_RUNTIME_SUBPATH)));
  }

  @Test
  void resolve_whenSnapWithCommonAndRuntimeUnwritable_expectCacheRuntimeFallback() {
    // Arrange
    Path root = tmp;
    Path home = root.resolve("home");
    Path common = root.resolve("snap-common");
    Path xdgCache = common.resolve(".cache");
    ensureDir(home);
    ensureDir(common);
    ensureDir(xdgCache);
    Map<String, String> env = new HashMap<>();
    env.put("SNAP", "/snap/app");
    env.put("UID", "99999");
    env.put("SNAP_INSTANCE_NAME", "cryptad.test");
    env.put("SNAP_USER_COMMON", common.toString());
    // Do not set XDG_RUNTIME_DIR; computeSnapRuntime will attempt /run (unwritable in tests) and
    // fall back to cache/rt.
    AppEnv ae = new AppEnv(env, "Linux", "user");
    AppDirs dirs = new AppDirs(env, sysProps(home, root), new HashMap<>(), ae);

    // Act
    Resolved r = dirs.resolve();

    // Assert
    assertTrue(r.configDir().startsWith(common.resolve("cryptad/config")));
    assertTrue(r.dataDir().startsWith(common.resolve("cryptad/data")));
    assertTrue(r.cacheDir().startsWith(xdgCache.resolve("cryptad")));
    assertTrue(r.runDir().startsWith(xdgCache.resolve("rt")));
    assertTrue(r.logsDir().startsWith(common.resolve("cryptad/logs")));
  }

  @Test
  void resolve_whenMacUsesXdg_expectLowercaseCryptadCasing() {
    // Arrange
    Path home = tmp.resolve("home");
    Path xdgConfig = tmp.resolve("xdg");
    ensureDir(home);
    ensureDir(xdgConfig);
    Map<String, String> env = new HashMap<>();
    env.put("XDG_CONFIG_HOME", xdgConfig.toString());
    AppEnv ae = new AppEnv(env, "Mac OS X", "user");
    Map<String, String> sp = sysProps(home, tmp);
    sp.put("os.name", "Mac OS X");
    AppDirs dirs = new AppDirs(env, sp, new HashMap<>(), ae);

    // Act
    Resolved r = dirs.resolve();

    // Assert
    assertTrue(norm(r.configDir().toString()).contains("/cryptad/config"));
  }

  @Test
  void resolve_whenWindowsAppDirs_expectCryptadCasing() {
    // Arrange
    Path root = tmp;
    Path home = root.resolve("home");
    Path roaming = home.resolve("AppData/Roaming");
    Path local = home.resolve("AppData/Local");
    ensureDir(roaming);
    ensureDir(local);
    Map<String, String> env = new HashMap<>();
    env.put("APPDATA", roaming.toString());
    env.put("LOCALAPPDATA", local.toString());
    AppEnv ae = new AppEnv(env, "Windows 10", "user");
    Map<String, String> sp = sysProps(home, root);
    sp.put("os.name", "Windows 10");
    AppDirs dirs = new AppDirs(env, sp, new HashMap<>(), ae);

    // Act
    Resolved r = dirs.resolve();

    // Assert
    assertTrue(norm(r.configDir().toString()).contains("/Cryptad/config"));
    assertTrue(norm(r.cacheDir().toString()).contains("/Cryptad"));
  }

  @Test
  void resolve_whenXdgRuntimeMissing_expectCacheRtFallback() {
    // Arrange
    Path root = tmp;
    Path home = root.resolve("home");
    Path xdgCache = root.resolve("xdg-cache");
    Path xdgConfig = root.resolve("xdg-config");
    Path xdgData = root.resolve("xdg-data");
    ensureDir(home);
    ensureDir(xdgCache);
    ensureDir(xdgConfig);
    ensureDir(xdgData);
    Map<String, String> env = new HashMap<>();
    env.put("XDG_CONFIG_HOME", xdgConfig.toString());
    env.put("XDG_DATA_HOME", xdgData.toString());
    env.put("XDG_CACHE_HOME", xdgCache.toString());
    AppEnv ae = new AppEnv(env, "Linux", "user");
    AppDirs dirs = new AppDirs(env, sysProps(home, root), new HashMap<>(), ae);

    // Act
    Resolved r = dirs.resolve();

    // Assert
    // Without XDG_RUNTIME_DIR and without a writable /run parent, runDir should be <cache>/rt
    assertTrue(r.runDir().startsWith(xdgCache.resolve("rt")));
  }

  @Test
  void resolve_whenFlatpakWithXdg_expectXdgAndFlatpakRuntime() {
    // Arrange
    Path root = tmp;
    Path home = root.resolve("home");
    Path xdgConfig = root.resolve("xdg-config");
    Map<String, String> env = getStringStringMap(root, home, xdgConfig);
    AppEnv ae = new AppEnv(env, "Linux", "user");
    AppDirs dirs = new AppDirs(env, sysProps(home, root), new HashMap<>(), ae);

    // Act
    Resolved r = dirs.resolve();

    // Assert
    assertTrue(r.configDir().startsWith(xdgConfig));
    assertTrue(
        norm(r.runDir().toString())
            .contains("/app/org.example.Cryptad/" + network.crypta.fs.Dirs.APP_RUNTIME_SUBPATH));
  }

  @NotNull
  private static Map<String, String> getStringStringMap(Path root, Path home, Path xdgConfig) {
    Path xdgData = root.resolve("xdg-data");
    Path xdgCache = root.resolve("xdg-cache");
    ensureDir(home);
    ensureDir(xdgConfig);
    ensureDir(xdgData);
    ensureDir(xdgCache);
    Map<String, String> env = new HashMap<>();
    env.put("FLATPAK_ID", "org.example.Cryptad");
    env.put("XDG_CONFIG_HOME", xdgConfig.toString());
    env.put("XDG_DATA_HOME", xdgData.toString());
    env.put("XDG_CACHE_HOME", xdgCache.toString());
    env.put("XDG_RUNTIME_DIR", root.resolve("xdg-rt").toString());
    return env;
  }

  @Test
  void expandAll_whenPlaceholderValuesPresent_expectResolvedDirectories() throws IOException {
    // Arrange
    Path root = tmp;
    Path home = root.resolve("home");
    ensureDir(home);
    Map<String, String> env = new HashMap<>();
    AppEnv ae = new AppEnv(env, "Linux", "tester", _ -> null);
    AppDirs dirs = new AppDirs(env, sysProps(home, root), new HashMap<>(), ae);
    Resolved r = dirs.resolve();
    String[] lines =
        new String[] {
          "node.install.cfgDir=${configDir}",
          "node.install.storeDir=stateDir", // leading token form
          "node.install.tempDir=cacheDir/tmp",
          "logger.dirname=${logsDir}",
          "End"
        };
    SimpleFieldSet sfs = new SimpleFieldSet(lines, true, true, false);

    // Act
    SimpleFieldSet out = CryptadConfig.expandAll(sfs, r, System.getProperties());

    // Assert
    String expandedConfigDir = out.get("node.install.cfgDir");
    String expandedTempDir = out.get("node.install.tempDir");
    System.out.println("DEBUG cfgDir expected=" + r.configDir());
    System.out.println("DEBUG node.install.cfgDir actual=" + expandedConfigDir);
    assertNotNull(expandedConfigDir);
    assertTrue(expandedConfigDir.startsWith(r.configDir().toString()));
    assertEquals(r.dataDir().toString(), out.get("node.install.storeDir"));
    assertNotNull(expandedTempDir);
    assertTrue(expandedTempDir.startsWith(r.cacheDir().toString()));
    System.out.println("DEBUG logsDir expected=" + r.logsDir());
    System.out.println("DEBUG logger.dirname actual=" + out.get("logger.dirname"));
    assertEquals(r.logsDir().toString(), out.get("logger.dirname"));
  }
}

File: src/test/kotlin/network/crypta/fs/AppDirsTest.kt
package network.crypta.fs

import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import network.crypta.config.expandAll
import network.crypta.support.SimpleFieldSet
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir

class AppDirsTest {
  @TempDir private var tmp: Path? = null

  private fun norm(value: String): String = value.replace('\\', '/')

  private fun tmpDirPath(): Path = requireNotNull(tmp)

  private fun sysProps(home: Path, tmpdir: Path): Map<String, String> {
    val props = HashMap<String, String>()
    props["user.home"] = home.toString()
    props["java.io.tmpdir"] = tmpdir.toString()
    props["os.name"] = "Linux"
    return props
  }

  @Test
  fun resolve_whenLinuxXdgUnset_defaultsUnderHome() {
    val home = tmpDirPath().resolve("home")
    val t = tmpDirPath().resolve("t")
    Files.createDirectories(home)
    Files.createDirectories(t)
    val env = HashMap<String, String>()
    val envResolver = AppEnv(env, "Linux", "tester")
    val dirs = AppDirs(env, sysProps(home, t), HashMap(), envResolver)

    val resolved = dirs.resolve()

    assertTrue(norm(resolved.configDir.toString()).contains(".config/cryptad/config"))
    assertTrue(norm(resolved.dataDir.toString()).contains(".local/share/cryptad/data"))
    assertTrue(Files.exists(resolved.configDir))
  }

  @Test
  fun resolve_whenLinuxXdgSet_respectsEnv() {
    val root = tmpDirPath()
    val home = root.resolve("home")
    val t = root.resolve("t")
    val xdgConfig = root.resolve("xdg-config")
    val xdgData = root.resolve("xdg-data")
    val xdgCache = root.resolve("xdg-cache")
    Files.createDirectories(home)
    Files.createDirectories(t)
    Files.createDirectories(xdgConfig)
    Files.createDirectories(xdgData)
    Files.createDirectories(xdgCache)
    val env = HashMap<String, String>()
    env["XDG_CONFIG_HOME"] = xdgConfig.toString()
    env["XDG_DATA_HOME"] = xdgData.toString()
    env["XDG_CACHE_HOME"] = xdgCache.toString()
    val envResolver = AppEnv(env, "Linux", "tester")
    val dirs = AppDirs(env, sysProps(home, t), HashMap(), envResolver)

    val resolved = dirs.resolve()

    assertTrue(resolved.configDir.startsWith(xdgConfig))
    assertTrue(resolved.dataDir.startsWith(xdgData))
    assertTrue(resolved.cacheDir.startsWith(xdgCache))
  }

  @Test
  fun resolve_whenMacNative_defaultsToLibrary() {
    val home = tmpDirPath().resolve("home")
    Files.createDirectories(home)
    val env = HashMap<String, String>()
    val envResolver = AppEnv(env, "Mac OS X", "user")
    val props = sysProps(home, tmpDirPath()).toMutableMap()
    props["os.name"] = "Mac OS X"
    val dirs = AppDirs(env, props, HashMap(), envResolver)

    val resolved = dirs.resolve()

    assertTrue(
      norm(resolved.configDir.toString()).contains("Library/Application Support/Cryptad/config")
    )
    assertTrue(norm(resolved.cacheDir.toString()).contains("Library/Caches/Cryptad"))
  }

  @Test
  fun resolve_whenMacXdgSet_respectsEnv() {
    val home = tmpDirPath().resolve("home")
    val xdgConfig = tmpDirPath().resolve("xdg")
    Files.createDirectories(home)
    Files.createDirectories(xdgConfig)
    val env = HashMap<String, String>()
    env["XDG_CONFIG_HOME"] = xdgConfig.toString()
    val envResolver = AppEnv(env, "Mac OS X", "user")
    val props = sysProps(home, tmpDirPath()).toMutableMap()
    props["os.name"] = "Mac OS X"
    val dirs = AppDirs(env, props, HashMap(), envResolver)

    val resolved = dirs.resolve()

    assertTrue(resolved.configDir.startsWith(xdgConfig))
  }

  @Test
  fun resolve_whenSnapStrict_usesCommonForData() {
    val root = tmpDirPath()
    val home = root.resolve("home")
    val common = root.resolve("snap-common")
    Files.createDirectories(home)
    Files.createDirectories(common)
    val env = HashMap<String, String>()
    env["SNAP"] = "/snap/app"
    env["SNAP_USER_COMMON"] = common.toString()
    env["SNAP_USER_DATA"] = home.toString()
    env["XDG_CONFIG_HOME"] = home.resolve(".config").toString()
    env["XDG_DATA_HOME"] = home.resolve(".local/share").toString()
    env["XDG_CACHE_HOME"] = home.resolve(".cache").toString()
    val envResolver = AppEnv(env, "Linux", "user")
    val dirs = AppDirs(env, sysProps(home, root), HashMap(), envResolver)

    val resolved = dirs.resolve()

    assertTrue(resolved.dataDir.startsWith(common.resolve("cryptad")))
  }

  @Test
  fun resolve_whenSnapWithoutCommon_usesXdgAndRuntimeUnderXdgRt() {
    val root = tmpDirPath()
    val home = root.resolve("home")
    val xdgConfig = root.resolve("xdg-config")
    val xdgData = root.resolve("xdg-data")
    val xdgCache = root.resolve("xdg-cache")
    val xdgRt = root.resolve("xdg-rt")
    Files.createDirectories(home)
    Files.createDirectories(xdgConfig)
    Files.createDirectories(xdgData)
    Files.createDirectories(xdgCache)
    Files.createDirectories(xdgRt)
    val env = HashMap<String, String>()
    env["SNAP"] = "/snap/app"
    env["XDG_CONFIG_HOME"] = xdgConfig.toString()
    env["XDG_DATA_HOME"] = xdgData.toString()
    env["XDG_CACHE_HOME"] = xdgCache.toString()
    env["XDG_RUNTIME_DIR"] = xdgRt.toString()
    val envResolver = AppEnv(env, "Linux", "user")
    val dirs = AppDirs(env, sysProps(home, root), HashMap(), envResolver)

    val resolved = dirs.resolve()

    assertTrue(resolved.configDir.startsWith(xdgConfig.resolve("cryptad/config")))
    assertTrue(resolved.dataDir.startsWith(xdgData.resolve("cryptad/data")))
    assertTrue(resolved.cacheDir.startsWith(xdgCache.resolve("cryptad")))
    assertTrue(resolved.runDir.startsWith(xdgRt.resolve(APP_RUNTIME_SUBPATH)))
  }

  @Test
  fun resolve_whenSnapWithCommonAndRuntimeUnwritable_fallsBackToCacheRuntime() {
    val root = tmpDirPath()
    val home = root.resolve("home")
    val common = root.resolve("snap-common")
    val xdgCache = common.resolve(".cache")
    Files.createDirectories(home)
    Files.createDirectories(common)
    Files.createDirectories(xdgCache)
    val env = HashMap<String, String>()
    env["SNAP"] = "/snap/app"
    env["UID"] = "99999"
    env["SNAP_INSTANCE_NAME"] = "cryptad.test"
    env["SNAP_USER_COMMON"] = common.toString()
    // Do not set XDG_RUNTIME_DIR; computeSnapRuntime will attempt /run (unwritable in tests) and
    // fall back to cache/rt.
    val envResolver = AppEnv(env, "Linux", "user")
    val dirs = AppDirs(env, sysProps(home, root), HashMap(), envResolver)

    val resolved = dirs.resolve()

    assertTrue(resolved.configDir.startsWith(common.resolve("cryptad/config")))
    assertTrue(resolved.dataDir.startsWith(common.resolve("cryptad/data")))
    assertTrue(resolved.cacheDir.startsWith(xdgCache.resolve("cryptad")))
    assertTrue(resolved.runDir.startsWith(xdgCache.resolve("rt")))
    assertTrue(resolved.logsDir.startsWith(common.resolve("cryptad/logs")))
  }

  @Test
  fun resolve_whenMacXdgCasing_isLowercaseCryptad() {
    val home = tmpDirPath().resolve("home")
    val xdgConfig = tmpDirPath().resolve("xdg")
    Files.createDirectories(home)
    Files.createDirectories(xdgConfig)
    val env = HashMap<String, String>()
    env["XDG_CONFIG_HOME"] = xdgConfig.toString()
    val envResolver = AppEnv(env, "Mac OS X", "user")
    val props = sysProps(home, tmpDirPath()).toMutableMap()
    props["os.name"] = "Mac OS X"
    val dirs = AppDirs(env, props, HashMap(), envResolver)

    val resolved = dirs.resolve()

    assertTrue(norm(resolved.configDir.toString()).contains("/cryptad/config"))
  }

  @Test
  fun resolve_whenWindowsAppDirs_casingIsCryptad() {
    val root = tmpDirPath()
    val home = root.resolve("home")
    val roaming = home.resolve("AppData/Roaming")
    val local = home.resolve("AppData/Local")
    Files.createDirectories(roaming)
    Files.createDirectories(local)
    val env = HashMap<String, String>()
    env["APPDATA"] = roaming.toString()
    env["LOCALAPPDATA"] = local.toString()
    val envResolver = AppEnv(env, "Windows 10", "user")
    val props = sysProps(home, root).toMutableMap()
    props["os.name"] = "Windows 10"
    val dirs = AppDirs(env, props, HashMap(), envResolver)

    val resolved = dirs.resolve()

    assertTrue(norm(resolved.configDir.toString()).contains("/Cryptad/config"))
    assertTrue(norm(resolved.cacheDir.toString()).contains("/Cryptad"))
  }

  @Test
  fun resolve_whenXdgRuntimeMissing_fallsBackToCacheRt() {
    val root = tmpDirPath()
    val home = root.resolve("home")
    val xdgCache = root.resolve("xdg-cache")
    val xdgConfig = root.resolve("xdg-config")
    val xdgData = root.resolve("xdg-data")
    Files.createDirectories(home)
    Files.createDirectories(xdgCache)
    Files.createDirectories(xdgConfig)
    Files.createDirectories(xdgData)
    val env = HashMap<String, String>()
    env["XDG_CONFIG_HOME"] = xdgConfig.toString()
    env["XDG_DATA_HOME"] = xdgData.toString()
    env["XDG_CACHE_HOME"] = xdgCache.toString()
    val envResolver = AppEnv(env, "Linux", "user")
    val dirs = AppDirs(env, sysProps(home, root), HashMap(), envResolver)

    val resolved = dirs.resolve()

    // Without XDG_RUNTIME_DIR and without a writable /run parent, runDir should be <cache>/rt
    assertTrue(resolved.runDir.startsWith(xdgCache.resolve("rt")))
  }

  @Test
  fun resolve_whenFlatpakXdgSet_usesXdgDirs() {
    val root = tmpDirPath()
    val home = root.resolve("home")
    val xdgConfig = root.resolve("xdg-config")
    val env = flatpakEnv(root, home, xdgConfig)
    val envResolver = AppEnv(env, "Linux", "user")
    val dirs = AppDirs(env, sysProps(home, root), HashMap(), envResolver)

    val resolved = dirs.resolve()

    assertTrue(resolved.configDir.startsWith(xdgConfig))
    assertTrue(
      norm(resolved.runDir.toString()).contains("/app/org.example.Cryptad/$APP_RUNTIME_SUBPATH")
    )
  }

  private fun flatpakEnv(root: Path, home: Path, xdgConfig: Path): Map<String, String> {
    val xdgData = root.resolve("xdg-data")
    val xdgCache = root.resolve("xdg-cache")
    Files.createDirectories(home)
    Files.createDirectories(xdgConfig)
    Files.createDirectories(xdgData)
    Files.createDirectories(xdgCache)
    val env = HashMap<String, String>()
    env["FLATPAK_ID"] = "org.example.Cryptad"
    env["XDG_CONFIG_HOME"] = xdgConfig.toString()
    env["XDG_DATA_HOME"] = xdgData.toString()
    env["XDG_CACHE_HOME"] = xdgCache.toString()
    env["XDG_RUNTIME_DIR"] = root.resolve("xdg-rt").toString()
    return env
  }

  @Test
  fun resolve_whenSystemdService_usesExportedDirs() {
    val root = tmpDirPath()
    val env = HashMap<String, String>()
    env["CONFIGURATION_DIRECTORY"] = root.resolve("etc").toString()
    env["STATE_DIRECTORY"] = root.resolve("lib").toString()
    env["CACHE_DIRECTORY"] = root.resolve("cache").toString()
    env["LOGS_DIRECTORY"] = root.resolve("log").toString()
    env["RUNTIME_DIRECTORY"] = root.resolve("run").toString()
    val svc = ServiceDirs(env, AppEnv(env, "Linux", "root"))

    val resolved = svc.resolve()

    assertTrue(resolved.configDir.startsWith(root.resolve("etc")))
    assertTrue(resolved.dataDir.startsWith(root.resolve("lib")))
    assertTrue(resolved.logsDir.startsWith(root.resolve("log")))
  }

  @Test
  fun resolve_whenWindowsService_rootsUnderProgramData() {
    val root = tmpDirPath()
    val env = HashMap<String, String>()
    env["PROGRAMDATA"] = root.resolve("ProgramData").toString()
    val svc = ServiceDirs(env, AppEnv(env, "Windows 10", "SYSTEM"))

    val resolved = svc.resolve()

    assertTrue(resolved.configDir.toString().contains("ProgramData"))
    assertTrue(resolved.logsDir.toString().contains("ProgramData"))
  }

  @Test
  fun resolve_whenMacDaemon_defaultsApply() {
    val env = HashMap<String, String>()
    val svc = ServiceDirs(env, AppEnv(env, "Mac OS X", "root"))

    val resolved = svc.resolve()

    assertTrue(
      norm(resolved.configDir.toString()).startsWith("/Library/Application Support/Cryptad/config")
    )
    assertTrue(norm(resolved.logsDir.toString()).startsWith("/Library/Logs/Cryptad"))
  }

  @Test
  @Throws(IOException::class)
  fun expandAll_whenPlaceholdersPresent_expandsAll() {
    val root = tmpDirPath()
    val home = root.resolve("home")
    Files.createDirectories(home)
    val env = HashMap<String, String>()
    val envResolver = AppEnv(env, "Linux", "tester") { null }
    val dirs = AppDirs(env, sysProps(home, root), HashMap(), envResolver)
    val resolved = dirs.resolve()
    val lines =
      arrayOf(
        $$"node.install.cfgDir=${configDir}",
        "node.install.storeDir=stateDir", // leading token form
        "node.install.tempDir=cacheDir/tmp",
        $$"logger.dirname=${logsDir}",
        "End",
      )
    val sfs = SimpleFieldSet(lines, true, true, false)

    val out = expandAll(sfs, resolved, System.getProperties())

    println("DEBUG cfgDir expected=${resolved.configDir}")
    println("DEBUG node.install.cfgDir actual=${out.get("node.install.cfgDir")}")
    assertTrue(out.get("node.install.cfgDir").startsWith(resolved.configDir.toString()))
    assertEquals(resolved.dataDir.toString(), out.get("node.install.storeDir"))
    assertTrue(out.get("node.install.tempDir").startsWith(resolved.cacheDir.toString()))
    println("DEBUG logsDir expected=${resolved.logsDir}")
    println("DEBUG logger.dirname actual=${out.get("logger.dirname")}")
    assertEquals(resolved.logsDir.toString(), out.get("logger.dirname"))
  }
}
